File: Joanna_Laurence/src/main/java/org/uva/sea/gui/widget/formatter/TextFormatterBuilder.java
package org.uva.sea.gui.widget.formatter;

import javafx.scene.control.TextFormatter;
import javafx.util.StringConverter;
import javafx.util.converter.*;
import org.uva.sea.gui.model.BaseQuestionModel;

import java.util.function.UnaryOperator;

public class TextFormatterBuilder<Value> {

    private StringConverter converter;
    private UnaryOperator<TextFormatter.Change> inputFilter;
    private Value defaultValue;
    private BaseQuestionModel questionModel;

    TextFormatterBuilder<Value> addDefaultValue(Value defaultValue) {
        this.defaultValue = defaultValue;
        return this;
    }

    TextFormatterBuilder<Value> addInputFilter(String pattern) {
        this.inputFilter = change -> {
            String text = change.getText();

            if (text.matches(pattern)) {
                return change;
            } else {
                System.out.println("Invalid input");
            }

            return null;
        };
        return this;
    }

    TextFormatterBuilder<Value> addBooleanConverter() {
        this.converter = new BooleanStringConverter();
        return this;
    }

    TextFormatterBuilder<Value> addDecimalConverter() {
        this.converter = new BigDecimalStringConverter();
        return this;
    }

    TextFormatterBuilder<Value> addIntegerConverter() {
        this.converter = new IntegerStringConverter();
        return this;
    }

    TextFormatterBuilder<Value> addDateConverter() {
        this.converter = new DateStringConverter();
        return this;
    }

    TextFormatterBuilder<Value> addMoneyConverter() {
        this.converter = new CurrencyStringConverter();
        return this;
    }

    TextFormatterBuilder<Value> addStringConverter() {
        this.converter = new ShortStringConverter();
        return this;
    }

    TextFormatter<Value> build() {
        return new TextFormatter<Value>(this.converter, this.defaultValue, this.inputFilter);
    }
}

File: niro/src/main/scala/nl/uva/se/sc/niro/ql/view/format/TextFormatterBuilder.scala
package nl.uva.se.sc.niro.ql.view.format

import java.util.function.UnaryOperator

import javafx.scene.control.TextFormatter
import javafx.util.StringConverter
import javafx.util.converter.{ BigDecimalStringConverter, BigIntegerStringConverter }

abstract class TextFormatterBuilder[T] {
  protected var converter: StringConverter[T] = _
  private var inputFilter: UnaryOperator[TextFormatter.Change] = _
  private var defaultValue: T = _

  def buildInputFilter(pattern: String): TextFormatterBuilder[T] = {
    inputFilter = (change: TextFormatter.Change) => if (change.getControlNewText.matches(pattern)) change else null
    this
  }

  def buildDefaultValue(defaultValue: T): TextFormatterBuilder[T] = {
    this.defaultValue = defaultValue
    this
  }

  def buildConverter(): TextFormatterBuilder[T]

  def build(): TextFormatter[T] =
    new TextFormatter[T](converter, defaultValue, inputFilter)
}
